### PR TITLE
Typescript migration

### LINKED ISSUE #125: Typescript migration discussion and notes
Continuing the discussion from #81

I needed someplace to talk about some of the stuff I'm running into.

So far it's just EsDoc and it's plugins are no longer maintained (haven't been touched since 2018), which means there's no hope of any work being done on the EsDoc typescript plugin. So I'll also have to migrate documentation to TypeDoc, which seams to be a good one-to-one replacement.

I'm currently loosely following my old [node-deno-template](https://github.com/josh-hemphill/node-deno-template).

I'll update as I run into more.

### PR DESCRIPTION
**Pull request by josh-hemphill**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
  - [x] Currently has some proof-of-concept style enforcement
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
  - [x] Most tests migrated and working, but some are unclear if they're old or unused. 
- [ ] All new and existing tests passed.

## Description

closes #125 - Typescript migration discussion and notes
related #114 - Missing typescript types
supersedes #81 - Migrate to typescript

Initial migration proof-of-concept.
Most things work already, and it has some lingering files for automating things I had copied from my repo I had used as a template. I'll get those working if there's interest.

The tests that are failing seem to be very difficult to diagnose, I'll probably need help with those. There's incorrect pixels...

### CODE REVIEW

File: .eslintcache [hunk 1]
```diff
@@ -0,0 +1 @@
...","severity":1,"message":"141","line":48,"column":13,"nodeType":"142","messageId":"143","endLine":48,"endColumn":24},{"ruleId":"140","severity":1,"message":"141","line":94,"column":96,"nodeType":"142","messageId":"143","endLine":94,"endColumn":109},{"ruleId":"144","severity":1,"message":"145","line":108,"column":104,"nodeType":"146","messageId":"147","endLine":108,"endColumn":105},{"ruleId":"140","severity":1,"message":"141","line":120,"column":23,"nodeType":"142","messageId":"143","endLine":120,"endColumn":36},{"ruleId":"140","severity":1,"message":"141","line":96,"column":17,"nodeType":"142","messageId":"143","endLine":96,"endColumn":28},{"ruleId":"140","severity":1,"message":"141","line":111,"column":17,"nodeType":"142","messageId":"143","endLine":111,"endColumn":30},"Parsing error: \"parserOptions.project\" has been set for @typescript-eslint/parser.\nThe file does not match your project config: mod.ts.\nThe file must be included in at least one of the projects provided.","no-console","Unexpected console statement.","MemberExpression","unexpected","@typescript-eslint/no-unused-vars","'e' is defined but never used. Allowed unused args must match /^_/u.","Identifier","unusedVar"]
```
Comment (joshmarinacci) [inline .eslintcache]: did you mean to include this file? This looks like something that shouldn't be in version control

**Comment by joshmarinacci** [comment]
This PR changes a whole lot of things, far too much for review. Could you make a PR with just the typescript changes, not all of the editor options and formatting changes (like adding semicolons to every line).

File: .eslintcache (hunk 1 above)
Comment (josh-hemphill) [reply .eslintcache]: Nope. thanks for catching that.

**Comment by josh-hemphill** [comment]
Yeah, there's a lot of stuff I grabbed from my template repo that I haven't cleaned up yet.
I had added linting because as I was rewriting things it wasn't clear if I should or shouldn't be using semicolons because it was all mixed depending on when it was written, so I just set it so it would all auto-format on save so I didn't have to worry about it.

I reverted trying to support Deno. It would require having to bundle all dependencies and then require Deno users to still add Node compatibility layer themselves because opentype.js and I think some of the other dependencies rely on Node.js built-in/standard modules.

**Comment by josh-hemphill** [comment]
I've removed most of the editor/formatting stuff. (only kept the `.editorconfig` since it's opt-in, has cross-editor support, and just informs line endings and indentation)

**Comment by joshmarinacci** [comment]
Did you intend to delete context.js and text.js

**Comment by josh-hemphill** [comment]
Yes. It's just replaced with a `context.ts` and `text.ts`, it didn't detect those as file name changes for some reason. 🤷

**Comment by joshmarinacci** [comment]
It looks like there are some conflicts to resolve before it can be merged.

**Comment by joshmarinacci** [comment]
Could you resolve the conflicts? Then we can try the merge again.

**Comment by josh-hemphill** [comment]
There's still 3 failing tests that I have no clue on.
And I probably still need to write a couple more to help with coving any type-checks that had to be added.

**Comment by josh-hemphill** [comment]
The individual commits now show the files being renamed and then edited, but merging only compares the final state to target branch, in which it considers the renamed files to be too considerably different and marks them as deletions and additions in the merge diff. If you want them to appear as renames, you'd have to use git rebase or merge follow options via the command line.

**Comment by joshmarinacci** [comment]
I've created a new port to typescript using ESBuild and Vitest, but I copied most of the actual typescript translation from your branch. Thank you so much. Please try out the new 0.4.* version with all of your hard work in them.

**Pull request closed without merging**